mitodl/open-discussions
# Added weekly/daily frontpage tasks

#### What are the relevant tickets?
Fixes #461 
Fixes #460 

#### What's this PR do?
Adds daily/weekly tasks for sending frontpage emails

#### How should this be manually tested?


##### Setup
- Ensure you valid valid settings in `.env` for the following settings: `MAILGUN_SENDER_DOMAIN`, `MAILGUN_KEY`, `MAILGUN_RECIPIENT_OVERRIDE`, `MAILGUN_RECIPIENT_OVERRIDE`, `OPEN_DISCUSSIONS_BASE_URL` (message me to discuss the necessary values)
- Also set `FEATURE_EMAIL_NOTIFICATIONS` and `FEATURE_FRONTPAGE_EMAIL_DIGESTS` to `True`
- Populate `NotificationSettings` for your user: `./manage.py populate_notification_settings --user USERNAME` (defaults to daily sending)
- Verify in a shell that only your user had a record for this (to avoid triggering a bunch of emails).
- Ensure your user has an email set and their profile has `email_optin=True` set.
- Ensure you have some posts visible to your user and some created today.

##### Testing
- First, test sending for users who have never received an email (default state when we release this feature): 
 ```python
from notifications import api
api.send_daily_frontpage_digests()
```
- Verify you now have an `EmailNotification` in pending status
- Then send the email:
```python
api.send_unsent_email_notifications()
```
- You should have received an email with valid links to the posts. There is minimal design for this in this PR.
- Run api.send_daily_frontpage_digests() again, you should not see an additional email.
- For the `EmailNotification` record, decrement `created_on` by 1 day and rerun `api.send_daily_frontpage_digests()`, you should see a new `EmailNotification` record and it should send ok.
- Repeat these steps with weekly sending if you wish, although the only code difference is the queries for `NotificationSettings` [here](https://github.com/mitodl/open-discussions/pull/504/files#diff-563bb4b2135d4c161c8fa3bee9e36919R59) vs [here](https://github.com/mitodl/open-discussions/pull/504/files#diff-563bb4b2135d4c161c8fa3bee9e36919R52).

#### Where should the reviewer start?

Look at notifications/apy.py, everything starts here:

- `send_{weekly,daily}_frontpage_emails` will eventually be called on a cron and these check every frontpage `NotificationSettings` for the desired frequency to see if there's any new posts to send
- If there are new posts since the last send, we create an `EmailNotification` in `pending` state.
- Then `send_unsent_email_notifications` (also eventually running on cron) picks these up, marks them and sending and triggers some batch tasks for them.
- Those batch tasks then execute, performing last-minute checks on the post content and render and send the email. This and `send_unsent_email_notifications` act as a reconciliation loop to ensure we eventually send all outstanding emails.

**_NOTE:_** this does introduce an inefficiency of having to look up the posts twice, but that is a tradeoff for delivery resiliency. Also, note in `notifications/utils.py` that we only attempt to send these emails once (so as to not spam the user if Mailgun fails, because if the mailgun api call fails we can't really tell if it was a full failure or not). We can always change this down the road.

#### Any background context you want to provide?

Things intentionally not done here (will deal with in later PRs):

- Add tasks to cron
